File: src/main/java/pr/eleks/we_at_her/services/api/ApiServiceFactory.java
package pr.eleks.we_at_her.services.api;

import org.springframework.stereotype.Service;
import pr.eleks.we_at_her.exceptions.UnknownServiceNameException;

import javax.annotation.PostConstruct;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class ApiServiceFactory {

    private List<ApiService> services;

    private static final Map<String, ApiService> myServiceCache = new HashMap<>();

    public ApiServiceFactory(List<ApiService> services) {
        this.services = services;
    }

    @PostConstruct
    public void initApiServiceCache() {
        for (ApiService service : services) {
            myServiceCache.put(service.getName(), service);
        }
    }

    public static ApiService getService(String name) throws UnknownServiceNameException {
        ApiService service = myServiceCache.get(name);
        if (service == null) throw new UnknownServiceNameException(name);
        return service;
    }
}

File: src/main/kotlin/pr/eleks/we_at_her/services/api/ApiServiceFactory.kt
package pr.eleks.we_at_her.services.api;

import org.springframework.stereotype.Service
import pr.eleks.we_at_her.exceptions.UnknownServiceNameException
import javax.annotation.PostConstruct

@Service
class ApiServiceFactory(
        private val services: List<ApiService>
){
    companion object{
        private val serviceCache: MutableMap<String, ApiService> = HashMap()

        @Throws(UnknownServiceNameException::class)
        fun getService(name: String): ApiService {
            return serviceCache[name]
                    ?: throw UnknownServiceNameException(name)
        }
    }

    @PostConstruct
    fun initApiServiceCache() {
        for (service in services) {
            serviceCache[service.getName()] = service
        }
    }
}
